README.md
# SensorOverlord

**Build:**
[![Build Status](https://travis-ci.com/ApfeldLab/SensorOverlord.svg?branch=master&status=passed)](https://travis-ci.com/github/ApfeldLab/SensorOverlord)
[![codecov](https://codecov.io/gh/apfeldlab/sensoroverlord/branch/master/graph/badge.svg)](https://codecov.io/gh/apfeldlab/sensoroverlord)

**Documentation:**
[![Docs](https://img.shields.io/badge/Documentation-Reference-green.svg)](https://apfeldlab.github.io/SensorOverlord/)


---------------------


This package aims to help you understand the limitations of quantatative measurements with two-state ratiometric sensors. 

You can load sensor spectra information into an object, take ratio measurements from that spectra, convert those ratio measurements into biophysically-meaningful values (e.g. fraction protenated and pH or fraction oxidized and redox potential), and then analyze how errors in ratio measurements affect the errors in those values. 

## Getting Started

You can install the package from within R with the `devtools` package. 

``` r
install.packages("devtools")
devtools::install_github("julianstanley/SensorOverlord")
```

You can then load the package like any other.

```r
library(sensorOverlord)
```

## Building a sensor

### Generic Sensor

You can build a sensor by one of two methods:

1. By using a spectra, usually provided as a .csv file of excitation-emission values

```r
# Build a spectra object from a csv file
spectra_file <- read.csv("spectra.csv", header = FALSE)
spectra <- spectraMatrixFromValues(
    lambdas_minimum = spectra_file,
    values_minimum = spectra_file,
    lambdas_maximum = spectra_file,
    values_maximum = spectra_file
)

# After building a spectra, I need to specify my ratiometric emission bands
# in this case, I'll specify (410 nm +/- 10 nm) / (470 nm +/- 10 nm)
my_sensor <- newSensorFromSpectra(spectra,
                     lambda_1 = c(400, 420),
                     lambda_2 = c(460, 480))
```


2. By using empirically-determined Rmin, Rmax, and delta values

```r
my_sensor <- new("Sensor", Rmin = 1, Rmax = 5, delta = 0.2)
```

### Specalized sensor

Once you have a generic sensor, you can create a more specalized sensor by passing
specalization-specific parameters.

For example, to make a redox sensor, I need to pass a midpoint potential, e0:

```r
my_redox_sensor <- new("redoxSensor", my_sensor, e0 = -250)
```

## Computing accurate ranges

One of the big outputs of this package is a plot that specifies tha accuracy 
of a given sensor at certain levels of microscopy precision.

To generate that output, you can call plotRanges() on a sensor object.

Right now, that range plot is only enabled for redoxSensors. For other sensors, 
hold tight (or, if you want them sooner, please email me at julianstanleya [at] gmail [dot] com 
and I'll majorly expedite the process!)

For example, here is how you would create that plot for the redox sensor created above,
across microscope inaccuracies ranging from 2% to 8% relative error, at "acceptable error"
thesholds of 1-5mV:

```r
library(sensorOverlord)

# Ranges plot of a sensor at different inaccuracies and thresholds
rangePlot(my_redox_sensor, ranges = ranges_df(my_redox_sensor,
                                        inaccuracies = seq(0.02, 0.08, by = 0.01),
                                        thresholds = 1:5))
```


NEWS.md
# sensorOverlord 0.2.0

* Computes error directly
* Extends computations to ligandSensor objects
* Adds rangePlot functionality
* Major improvements and UI redesign of shiny application

# sensorOverlord 0.1.0

* First functional version of the package, published alongwide biorxiv preprint.

inst/shinyApp/ui.R
require(shiny)
require(shinydashboard)
require(sensorOverlord)
require(dplyr)
require(ggplot2)
require(ggalt)
require(mongolite)
source("helpers.R")

#' @import shinydashboard


welcomeMessage <- "
    Welcome to the SensorOverlord home page. Please select a sensor from
    our database, or go to 'Add a Custom Sensor' to input parameters of
    your own.
"

# Sidebar definition ------------------------------------------------------
sidebar <- dashboardSidebar(
    sidebarMenu(
        menuItem("Home", tabName = "home"),
        menuItem("Add a Custom Sensor", tabName = "newSensor"),
        menuItem("Full Error Table", tabName = "table"),
        menuItem("Settings", tabName = "settings"),
        menuItem("Browse Sensor Database", tabName = "browse"),
        menuItem("Contribute to Database", tabName = "upload"))
)



# Body definition --------------------------------------------------------------
body <- dashboardBody(
    tabItems(
        # Home Page ------------------------------------------------------------
        tabItem(tabName = "home",
                    # Welcome message
                    h3(welcomeMessage),
                    br(),

                    fluidRow(
                        box(
                        align = 'left', width = 12,
                        title = "Choose a sensor and error model",

                        # Select a sensor
                        selectInput(
                            inputId = "sensors",
                            label = "Select a sensor",
                            choices = c("Custom (see 'Add a Custom Sensor')", sensorNames),
                            selected = sensorNames[1],
                            multiple = FALSE
                            ),

                        # Input a relative error
                        numericInput(
                                inputId = "relErr",
                                label = "Relative Error",
                                min = 0,
                                max = Inf,
                                step = 0.01,
                                value = 0.01
                            ),

                        # Input an absolute error
                        numericInput(
                                inputId = "absErr",
                                label = "Absolute error",
                                min = 0,
                                max = Inf,
                                step = 0.01,
                                value = 0
                            ),

                        # Input a desired accuracy
                        numericInput(
                                inputId = "acc",
                                label = "Desired accuracy",
                                min = 0,
                                max = Inf,
                                step = 0.01,
                                value = 2
                            )


                        )
                    ),

                    # Plot the suitable ranges
                fluidRow(
                    box(
                        h3("Suitable range of selected sensor:",
                           align = "center"),
                        h3(textOutput("rangeText"),
                           align = "center")
                    ),

                    box(
                        h3("Sensor characteristics:",
                           align = "center"),
                        h3(textOutput("sensorChars"),
                           align = "center")
                    )
                ),

                    br(),

                fluidRow(
                    box(
                        plotOutput(outputId = "range")
                    ),

                    box(
                        plotOutput(outputId = "phasePlot")
                    )
                )

            # Close overview tab
            ),

        # Custom Sensor Page ---------------------------------------------------
        tabItem(tabName = "newSensor",
                # Sensor characteristics selection
                box(
                    title = "Input sensor characteristics",

                    # Upload a file
                    fileInput(
                        inputId = "customSpectra",
                        label = "Upload spectra as .csv"
                    ),

                    # Type of sensor
                    radioButtons(
                        inputId = "sensorType",
                        label = "Sensor Type",
                        choiceValues = c("redox", "pH", "other"),
                        choiceNames = c("Redox", "pH", "Other"),
                    ),

                    # Rmin input
                    numericInput(
                        inputId = "Rmin",
                        label = "Rmin",
                        min = 0,
                        max = Inf,
                        step = 0.01,
                        value = 1
                    ),

                    # Rmax input
                    numericInput(
                        inputId = "Rmax",
                        label = "Rmax",
                        min = 0,
                        max = Inf,
                        step = 0.01,
                        value = 5
                    ),

                    # Delta input
                    numericInput(
                        inputId = "delta",
                        label = withMathJax("$$ \\delta_{\\lambda 2} $$"),
                        min = 0,
                        max = Inf,
                        step = 0.01,
                        value = 0.2
                    ),

                    # Midpoint selection
                    numericInput(
                        inputId = "midpoint",
                        label = "Midpoint Value (E0, pKa, etc)
                                \n Note: Ignored for 'other' sensor type",
                        min = -Inf,
                        max = Inf,
                        step = 1,
                        value = 1
                    )
                ),

                # Sensor characteristics output
                box(
                    h3(textOutput("customChars")),
                    br(),
                    plotOutput("plotFractionMax_custom"),
                    plotOutput("plotValue_custom")
        )
        ),



        # Full Error Table Page-------------------------------------------------
        tabItem(tabName = "table",
                h1("Full error table for the selected sensor"),
                dataTableOutput('fullTable')
                ),

        # Settings Page --------------------------------------------------------
        tabItem(tabName = "settings",
                h1("Settings!"),

                # Input boxes
                fluidRow(
                    box(
                        title = "Generating ratiometric intensity values",

                        numericInput(
                            inputId = "rpres",
                            label = "Distance between R values (log scale)",
                            min = -Inf,
                            max = Inf,
                            step = 0.01,
                            value = -2
                        ),

                        numericInput(
                            inputId = "rpres_edge",
                            label = "Fold-increase in distance between R values
                            near Rmin and Rmax (log scale)",
                            min = -Inf,
                            max = Inf,
                            step = 0.01,
                            value = 2
                        )
                    )

                ),

                fluidRow(
                    box(align="center",
                        width = 12,
                        h3(textOutput(outputId = "numR")),
                        h3(textOutput(outputId = "precision"))
                    ),

                    box(align="left",
                        width = 6,
                        plotOutput(outputId = 'numRHist')),
                    box(align="right",
                        width = 6,
                        plotOutput(outputId = "precisionHist"))
                )



        )


    # Close tabItems and dashboardBody ----------------------------------------
    )
)


# Putting the page components together ----------------------------------------
dashboardPage(skin = "blue",
    # Header
    dashboardHeader(title = "Sensor Overlord"),
    sidebar,
    body,
    title = "Sensor Overlord"
)


